CMakeLists.txt

```diff
@@ -1370,27 +1370,56 @@ if(UNIX AND NOT APPLE)
   )
 
   option(INSTALL_USER_UDEV_RULES "Install user udev rule file for USB HID and Bulk controllers" ON)
-  if (INSTALL_USER_UDEV_RULES)
-    install(
-      FILES
-        "${CMAKE_CURRENT_SOURCE_DIR}/res/linux/mixxx-usb-uaccess.rules"
-      DESTINATION
-        "${MIXXX_INSTALL_DATADIR}/udev/rules.d"
-    )
-    install(CODE "
-    message(STATUS \"Important Note: Installation of udev rules\n\"
-        \"The udev rule file for USB HID and Bulk controller permissions will be\n\"
-        \"installed to:\n\"
-        \"    ${CMAKE_INSTALL_PREFIX}/${MIXXX_INSTALL_DATADIR}/udev/rules.d.\n\"
-        \"If you are installing Mixxx from source for your own use, copy\n\"
-        \"mixxx-usb-uaccess.rules to /etc/udev/rules.d/ and run:\n\"
-        \"    udevadm control --reload-rules && udevadm trigger\n\"
-        \"as root to load the rules.\n\"
-        \"If you are building a package for a distribution, the correct directory for\n\"
-        \"system rules is either /lib/udev/rules.d (e.g. Debian, Fedora) or\n\"
-        \"/usr/lib/udev/rules.d (e.g. Arch Linux) with an appropriate priority prefix.\n\"
-        \"Adjust your package script accordingly and set -DINSTALL_USER_UDEV_RULES=OFF\")
-    ")
+  if(INSTALL_USER_UDEV_RULES)
+    set(MIXXX_UDEVDIR "${MIXXX_INSTALL_DATADIR}/udev")
+    if (CMAKE_INSTALL_PREFIX STREQUAL "/usr")
+      message(FATAL_ERROR "Bla")
+      find_package(PkgConfig)
+      if (PKG_CONFIG_FOUND)
+        pkg_check_modules( PKGCONFIG_UDEV udev)
+        if (PKGCONFIG_UDEV_FOUND)
+	  execute_process(
+	    COMMAND ${PKG_CONFIG_EXECUTABLE} --variable=udevdir udev
+            OUTPUT_VARIABLE PKGCONFIG_UDEVDIR
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+          )
+          if(PKGCONFIG_UDEVDIR)
+	    file(TO_CMAKE_PATH "${PKGCONFIG_UDEVDIR}" MIXXX_UDEVDIR)
+          endif()
```

Comment:
Are you sure that PKGCONFIG_UDEVDIR is not an absolute path outside CMAKE_INSTALL_PREFIX? Mixxx mustn't install anything outside the chose prefix under any circumstances.

Reply:
It does not:
```
-- Installing: /home/be/local/share/mixxx/udev/rules.d/mixxx-usb-uaccess.rules
```

Reply:
And this is certainly the case on all systems? So returning a relative path instead of an absolute one is guaranteed by pkg-config?

Reply:
I don't know.

Reply:
> Are you sure that PKGCONFIG_UDEVDIR is not an absolute path outside CMAKE_INSTALL_PREFIX? Mixxx mustn't install anything outside the chosen prefix under any circumstances.

In case of udevdir it is an exception. The rules files have to be installed at the common folder returned by pkg-config IF Mixxx is installed in the system location "/usr". This should happens only in the packaging case. 

In all other case the udevdir is useless and the rules file is installed under our fall back location .../share/mixxx/udev/rules.d/mixxx-usb-uaccess.rules as before. 

This should make all hacks in the packaging scripts obsolete. 

Do we know a distro where this assumption is void?

Reply:
> In all other case the udevdir is useless and the rules file is installed under our fall back location .../share/mixxx/udev/rules.d/mixxx-usb-uaccess.rules as before.

It seems the [rules file](https://man.archlinux.org/man/udev.7#RULES_FILES) can also be installed to `/usr/local/lib/udev/rules.d/` in Arch Linux's case, but this doesn't seem to hold on e.g. Debian (I just checked the recognized locations using `man 7 udev` on an old Debian host).

I guess the fallback is the only thing that makes sense in that situation unfortunately :(

Reply:
Sorry, but there is no exception to that rule. If the uservs system doesn't support loading udev rules from `/usr/local`, it's the user's responsibility to move the file afterwards or choose a different prefix.

Reply:
Ubuntu the system install prefix is /usr the udev path is /lib/udev, that is an exception weakening your rule.

It is just that the pkg-config udefdir returns the system udev dir where we want to install our rule if Mixxx is installed to the system path.
If this is outside of /usr we need to respect it anyway. 
I think it is Ok to trust pkg-config here

Reply:
> Ubuntu the system install prefix is /usr the udev path is /lib/udev, that is an exception, weakening your rule.

If you want your udev rules to be placed into `/lib/udev` instead of `/usr/lib/udev`, then use `-DCMAKE_INSTALL_PREFIX=/`. In that case, it's okay to use the output from pkg config to find the udev path, because there is no limitation on what path we may write to. In all other cases it needs to be certain that deleting mixxx files from `CMAKE_INSTALL_PREFIX` leaves no traces of mixxx and I don't need to go look anywhere else because mixxx just decided to ignore my CMAKE_INSTALL_PREFIX.

Reply:
That would be wrong for Ubuntu sorry. Look into our rebian package.

Reply:
This rule applies to Arch, but it is not on us to force that to any distro. 

What is the alternative?

For my impression I use the udevdir in PC for what I this for.

The package maintainers install Mixxx to the location where it should be "/usr" and we use the  udev path PC gives us.

Reply:
> That would be wrong for Ubuntu sorry

What would be wrong? Have you read the special case section at: https://cmake.org/cmake/help/v3.15/module/GNUInstallDirs.html?highlight=gnuinstalldirs#special-cases

> The following values of CMAKE_INSTALL_PREFIX are special:
>
> - `/`
>   For `<dir>` other than the `SYSCONFDIR`, `LOCALSTATEDIR` and `RUNSTATEDIR`, the value of `CMAKE_INSTALL_<dir>` is prefixed with `usr/` if it is not user-specified as an absolute path. For example, the INCLUDEDIR value include becomes usr/include. This is required by the GNU Coding Standards, which state:
>   When building the complete GNU system, the prefix will be empty and /usr will be a symbolic link to /.

Reply:
Ah Ok, so we need to check how debuild uses CMake

Reply:
What does PC return for udevdir on arch?

Reply:
```
$ pkgconf --variable=udevdir udev
/usr/lib/udev
```

Arch does not use `/lib` or `/lib64` (it is symlinked to `/usr/lib`), as well as we also don't do `/usr/sbin` or `/bin` or `/sbin` (it is symlinked to `/usr/bin`): https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/

Reply:
Cool, so it should work for all major distros at least